src/rpm/common.go

```diff
@@ -0,0 +1,281 @@
+
+type rpmTagType struct {
+	ID    int
+	Type  rpmValueType
+	Value interface{}
+}
+
+type rpmTagSetType []rpmTagType
+
+type packedTagType struct {
+	Count int
+	Data  *bytes.Buffer
+}
+
+func (tagSet *rpmTagSetType) addTags(tags ...rpmTagType) {
+	*tagSet = append(*tagSet, tags...)
+}
+
+func packValues(values ...interface{}) *bytes.Buffer {
+	buf := bytes.NewBuffer(nil)
+
+	for _, v := range values {
+		binary.Write(buf, binary.BigEndian, v)
+	}
+
+	return buf
+}
+
+func packTag(tag rpmTagType) (*packedTagType, error) {
+	var packed packedTagType
+	packed.Data = bytes.NewBuffer(nil)
+
+	switch tag.Type {
+	case rpmTypeNull: // NULL
+		// XXX: It should be array of nil's ??
+	case rpmTypeChar: // CHAR
+		// XXX: It should be array of rune's or bytes ??
+	case rpmTypeBin: // BIN
+		byteArray, ok := tag.Value.([]byte)
+		if !ok {
+			return nil, fmt.Errorf("BIN value should be []byte")
+		}
+
+		packed.Count = len(byteArray)
+		for _, byteValue := range byteArray {
+			if _, err := io.Copy(packed.Data, packValues(byteValue)); err != nil {
+				return nil, err
+			}
+		}
+	case rpmTypeStringArray: // STRING_ARRAY
+		// value should be strings array
+		stringsArray, ok := tag.Value.([]string)
+		if !ok {
+			return nil, fmt.Errorf("STRING_ARRAY value should be []string")
+		}
+
+		packed.Count = len(stringsArray)
+
+		for _, v := range stringsArray {
+			bytedString := []byte(v)
+			bytedString = append(bytedString, 0)
+			if _, err := io.Copy(packed.Data, packValues(bytedString)); err != nil {
+				return nil, err
+			}
+		}
+	case rpmTypeString: // STRING
+		// value should be string
+		stringValue, ok := tag.Value.(string)
+		if !ok {
+			return nil, fmt.Errorf("STRING value should be string")
+		}
+
+		packed.Count = 1
+
+		bytedString := []byte(stringValue)
+		bytedString = append(bytedString, 0)
+		if _, err := io.Copy(packed.Data, packValues(bytedString)); err != nil {
+			return nil, err
+		}
+
+	case rpmTypeInt8: // INT8
+		// value should be []int8
+		int8Values, ok := tag.Value.([]int8)
+		if !ok {
+			return nil, fmt.Errorf("INT8 value should be []int8")
+		}
+
+		packed.Count = len(int8Values)
+
+		if _, err := io.Copy(packed.Data, packValues(int8Values)); err != nil {
+			return nil, err
+		}
+
+	case rpmTypeInt16: // INT16
+		// value should be []int16
+		int16Values, ok := tag.Value.([]int16)
+		if !ok {
+			return nil, fmt.Errorf("INT16 value should be []int16")
+		}
+
+		packed.Count = len(int16Values)
+
+		for _, value := range int16Values {
+			if _, err := io.Copy(packed.Data, packValues(value)); err != nil {
+				return nil, err
+			}
+		}
+
+	case rpmTypeInt32: // INT32
+		// value should be []int32
+		int32Values, ok := tag.Value.([]int32)
+		if !ok {
+			return nil, fmt.Errorf("INT32 value should be []int32")
+		}
+
+		packed.Count = len(int32Values)
+
+		for _, value := range int32Values {
+			if _, err := io.Copy(packed.Data, packValues(value)); err != nil {
+				return nil, err
+			}
+		}
+
+	case rpmTypeInt64: // INT64
```

Comment:
Are all this types really used?

Reply:
No, but I think it's better ho have a code that works for all types now.

Reply:
The main problem of such approach - use have unused and untested code...